File: days/day_02/solutions/solution.ts
import * as fs from 'fs';
import * as path from 'path';

// Check if ID is invalid for Part 1: exactly two identical sequences
function isInvalidPart1(idStr: string): boolean {
    const n = idStr.length;
    if (n % 2 !== 0) {
        return false;
    }
    const half = n / 2;
    return idStr.slice(0, half) === idStr.slice(half);
}

// Check if ID is invalid for Part 2: sequence repeated 2+ times
function isInvalidPart2(idStr: string): boolean {
    const n = idStr.length;
    for (let k = 2; k <= n; k++) {
        if (n % k === 0) {
            const seqLen = n / k;
            const pattern = idStr.slice(0, seqLen);
            const repeated = pattern.repeat(k);
            if (idStr === repeated) {
                return true;
            }
        }
    }
    return false;
}

// Parse a range string like "start-end"
function parseRange(rangeStr: string): [number, number] {
    const parts = rangeStr.split('-');
    const start = parseInt(parts[0].trim(), 10);
    const end = parseInt(parts[1].trim(), 10);
    return [start, end];
}

// Parse a line of comma-separated ranges
function parseRanges(line: string): Array<[number, number]> {
    return line.split(',')
        .map(s => s.trim())
        .filter(s => s.length > 0)
        .map(parseRange);
}

function solve(inputData: string): [string, string] {
    const lines = inputData.trim().split('\n');
    
    let part1Sum = 0;
    let part2Sum = 0;
    
    for (const line of lines) {
        if (line.trim() === '') {
            continue;
        }
        const ranges = parseRanges(line);
        
        for (const [start, end] of ranges) {
            for (let num = start; num <= end; num++) {
                const idStr = num.toString();
                
                if (isInvalidPart1(idStr)) {
                    part1Sum += num;
                }
                
                if (isInvalidPart2(idStr)) {
                    part2Sum += num;
                }
            }
        }
    }
    
    return [part1Sum.toString(), part2Sum.toString()];
}

function main(): void {
    const inputPath = path.join(__dirname, '../data/input.txt');
    const data = fs.readFileSync(inputPath, 'utf-8');
    const [part1, part2] = solve(data);
    console.log(`Part 1: ${part1}`);
    console.log(`Part 2: ${part2}`);
}

main();

File: days/day_02/solutions/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

class Solution {
    // Check if ID is invalid for Part 1: exactly two identical sequences
    private static boolean isInvalidPart1(String idStr) {
        int n = idStr.length();
        if (n % 2 != 0) {
            return false;
        }
        int half = n / 2;
        return idStr.substring(0, half).equals(idStr.substring(half));
    }
    
    // Check if ID is invalid for Part 2: sequence repeated 2+ times
    private static boolean isInvalidPart2(String idStr) {
        int n = idStr.length();
        for (int k = 2; k <= n; k++) {
            if (n % k == 0) {
                int seqLen = n / k;
                String pattern = idStr.substring(0, seqLen);
                String repeated = pattern.repeat(k);
                if (idStr.equals(repeated)) {
                    return true;
                }
            }
        }
        return false;
    }
    
    // Parse a range string like "start-end"
    private static long[] parseRange(String rangeStr) {
        String[] parts = rangeStr.split("-");
        long start = Long.parseLong(parts[0].trim());
        long end = Long.parseLong(parts[1].trim());
        return new long[]{start, end};
    }
    
    // Parse a line of comma-separated ranges
    private static List<long[]> parseRanges(String line) {
        List<long[]> ranges = new ArrayList<>();
        String[] parts = line.split(",");
        for (String part : parts) {
            part = part.trim();
            if (!part.isEmpty()) {
                ranges.add(parseRange(part));
            }
        }
        return ranges;
    }
    
    public static String[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        
        long part1Sum = 0;
        long part2Sum = 0;
        
        for (String line : lines) {
            if (line.trim().isEmpty()) {
                continue;
            }
            List<long[]> ranges = parseRanges(line);
            
            for (long[] range : ranges) {
                long start = range[0];
                long end = range[1];
                for (long num = start; num <= end; num++) {
                    String idStr = String.valueOf(num);
                    
                    if (isInvalidPart1(idStr)) {
                        part1Sum += num;
                    }
                    
                    if (isInvalidPart2(idStr)) {
                        part2Sum += num;
                    }
                }
            }
        }
        
        return new String[]{String.valueOf(part1Sum), String.valueOf(part2Sum)};
    }
    
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String data = Files.readString(inputPath);
        String[] results = solve(data);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
}
